src/util/address.rs

```diff
@@ -523,23 +533,23 @@ impl Address {
 
     /// Get the address type of the address.
     /// None if unknown, non-standard or related to the future witness version.
-    pub fn address_type(&self) -> Option<AddressType> {
+    pub fn address_type(&self) -> Result<AddressType, Error> {
         match self.payload {
-            Payload::PubkeyHash(_) => Some(AddressType::P2pkh),
-            Payload::ScriptHash(_) => Some(AddressType::P2sh),
+            Payload::PubkeyHash(_) => Ok(AddressType::P2pkh),
+            Payload::ScriptHash(_) => Ok(AddressType::P2sh),
             Payload::WitnessProgram {
                 version,
                 program: ref prog,
             } => {
                 // BIP-141 p2wpkh or p2wsh addresses.
                 match version {
                     WitnessVersion::V0 => match prog.len() {
-                        20 => Some(AddressType::P2wpkh),
-                        32 => Some(AddressType::P2wsh),
-                        _ => None,
+                        20 => Ok(AddressType::P2wpkh),
+                        32 => Ok(AddressType::P2wsh),
+                        len => Err(Error::InvalidWitnessProgramLength(len)),
                     },
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                    WitnessVersion::V1 => Ok(AddressType::P2tr),
```

Comment:
we still need the `if prog.len() == 32` check

Reply:
Sorry, I am not getting why if you just pointed out that this is not a part of the spec

Reply:
Ok, got it. If we will have non-32-byte V1 it will be _non-taproot_ V1